### PR TITLE
bugfixes, ntriangles for higher order triangle and introduction of new kwarg `copy_fields` in `interpolategradient()`

### PR DESCRIPTION
**Pull request by JanM12**
- fixed some bugs in `src/makieploting.jl`
-  changed parametrization for ntriangles for higher order triangle (`ntriangles(cell::Ferrite.AbstractCell{2,3,3})  = 1` -> `ntriangles(cell::Ferrite.AbstractCell{2,N,3}) where {N} = 1`)
- introduced kwarg `copy_fields` to interpolategradient() which copies all selected fields of the DofHandler and all values of the node vector passed to the function to the returned DofHandler and and node vector. This way one can plot a gradient field in deformed state for example.

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -177,8 +177,8 @@ num_vertices(p::MakiePlotter) = size(p.physical_coords,1)
 TODO this looks faulty...think harder.
 """
 # Helper to count triangles e.g. for preallocations.
-ntriangles(cell::Ferrite.AbstractCell{2,3,3}) = 1 # Tris in 2D
-ntriangles(cell::Ferrite.AbstractCell{3,3,1}) = 1 # Tris in 3D
+ntriangles(cell::Ferrite.AbstractCell{2,N,3}) where {N} = 1 # Tris in 2D
+ntriangles(cell::Ferrite.AbstractCell{3,N,1}) where {N} = 1 # Tris in 3D
```
Comment (koehlerson) [inline src/utils.jl:181]:
```suggestion
ntriangles(cell::Ferrite.AbstractCell{3,3,1}) where {N} = 1 # Tris in 3D
```

File: src/makieplotting.jl [hunk 2]
```diff
@@ -172,9 +172,10 @@ function Makie.plot!(WF::Wireframe{<:Tuple{<:MakiePlotter{dim}}}) where dim
     # u_matrix = @lift($(WF[:deformation_field])===:default ? zeros(0,3) : transfer_solution(plotter; field_idx=Ferrite.find_field(plotter.dh,$(WF[:deformation_field])), process=identity))
     # coords = @lift($(WF[:deformation_field])===:default ? plotter.physical_coords : plotter.physical_coords .+ ($(WF[:scale]) .* $(u_matrix)))
     #original representation
+    pointtype = dim > 2 ? Point3f : Point2f
```
Comment (termi-official) [inline src/makieplotting.jl:175]: Maybe we can be more specific regarding the dimension, so we can e.g. error in the case dim = 1 or dim > 3.

Comment (termi-official) [reply src/makieplotting.jl:175]: or get the type via some dispatch

File: src/utils.jl [hunk 3]
```diff
@@ -440,7 +440,7 @@ This is a helper to access the correct value in Tensors.jl entities, because the
 
 Compute the piecewise discontinuous gradient field for `field_name`. Returns the flux dof handler and the corresponding flux dof values.
 """
-function interpolate_gradient_field(dh::Ferrite.DofHandler{spatial_dim}, u::AbstractVector, field_name::Symbol) where {spatial_dim}
+function interpolate_gradient_field(dh::Ferrite.DofHandler{spatial_dim}, u::AbstractVector, field_name::Symbol; copy_fields=Symbol[]) where {spatial_dim}
```
Comment (termi-official) [inline src/utils.jl:443]: Docstring

**Review by termi-official — CHANGES_REQUESTED** [review]
Sorry I cannot find the `ntriangles` for higher order part. Is it missing?

File: src/utils.jl [hunk 4]
```diff
@@ -177,7 +177,7 @@ num_vertices(p::MakiePlotter) = size(p.physical_coords,1)
 TODO this looks faulty...think harder.
 """
 # Helper to count triangles e.g. for preallocations.
-ntriangles(cell::Ferrite.AbstractCell{2,3,3}) = 1 # Tris in 2D
+ntriangles(cell::Ferrite.AbstractCell{2,N,3}) where {N} = 1 # Tris in 2D
```
[the remark this answers is not in the document]
Comment (termi-official) [reply src/utils.jl:180]: Okay seems like Github has some caching issues currently, this was not visible in the review before you commented. :D

File: src/makieplotting.jl (hunk 2 above)
Comment (koehlerson) [reply src/makieplotting.jl:175]: But there are only these two types. What's the benefit of a dispatch at this case?

Comment (termi-official) [reply src/makieplotting.jl:175]: Benefit of dispatch over having the error management via if/else? None. Benefit over having no error management? Clearer error messages for users.

Comment (koehlerson) [reply src/makieplotting.jl:175]: And on what you want to dispatch? A Val? It's unclear to me and Jan how it should be changed

Comment (termi-official) [reply src/makieplotting.jl:175]: Maybe something like
```suggestion
    if dim == 2
        pointtype = Point2f
    elseif dim == 3
        pointtype = Point3f
    else
       @error "Dimension $dim not supported for wireframe"
    end
```
or 
```suggestion
getpointtype(::Val{3}) = Point3f
...
```
not sure what works best here. Having dispatches here reduces code duplication (we have such ifs at several points in the code). Also keeping https://github.com/Ferrite-FEM/FerriteViz.jl/issues/41 in the back of my head here.

Comment (koehlerson) [reply src/makieplotting.jl:175]: Yes #41 needs 2D as well right, that's why we wrote it initially like this, but then let's do the dispatch approach @JanM12

Comment (termi-official) [reply src/makieplotting.jl:175]: Depends on your use case. I think it is also nice to have potting routines for pure 1D problems.

Comment (koehlerson) [reply src/makieplotting.jl:175]: Yeah but they need to be plotted on a 2D surface right :D x for the domain and y for the data I guess

File: src/makieplotting.jl [hunk 5]
```diff
@@ -227,7 +228,7 @@ function Makie.plot!(WF::Wireframe{<:Tuple{<:MakiePlotter{dim}}}) where dim
     cellset_u =  reshape(transfer_scalar_celldata(plotter, cellset_u; process=identity), num_vertices(plotter))
     Makie.mesh!(WF, plotter.mesh, color=cellset_u, shading=false, scale_plot=false, colormap=:darktest, visible=WF[:cellsets])
     #plot the nodes
-    Makie.scatter!(WF,gridnodes,markersize=WF[:markersize], color=WF[:color], visible=WF[:visible])
+    Makie.scatter!(WF,gridnodes,markersize=WF[:markersize], color=WF[:color], visible=WF[:plotnodes])
```
Comment (termi-official) [inline src/makieplotting.jl:231]: Shouldn't this be something like `@lift WF[:visible] && WF[:plotnodes]` such that this one is only rendered when the wireframe is visible?

**Pull request merged**